File: Delphi/Modules/25.Printing and Exporting/20.CustomPreview/UProgressThread.pas
unit UProgressThread;

interface
uses Classes, SysUtils, VCL.FlexCel.Core;
type

TProgressThread = class(TThread)
  protected
    ProgressFeedback: TProc<integer, string>;
    FinalFeedback: TProc<boolean, string>;
    Xls: TExcelFile;
    FileName: string;
    AllSheets: boolean;
    Canceled: boolean;

{$IF CompilerVersion > 22} //Delphi XE doesn't support TerminatedSet, so sadly in XE we will have to call it manually.
    procedure TerminatedSet; override;
{$ifend}
    procedure DoTerminate; override;
  public
    constructor Create(const aXls: TExcelFile; const aProgressFeedback: TProc<integer, string>;
    const aFinalFeedback: TProc<boolean, string>; const aFileName: string; const aAllSheets: boolean);
{$IF CompilerVersion <= 22} //Delphi XE doesn't support TerminatedSet, so sadly in XE we will have to call it manually.
     procedure Cancel;
{$ifend}
end;

implementation

{ TProgressThread }

constructor TProgressThread.Create(const aXls: TExcelFile;
  const aProgressFeedback: TProc<integer, string>;
  const aFinalFeedback: TProc<boolean, string>; const aFileName: string; const aAllSheets: boolean);

begin
  inherited Create(True);
  FreeOnTerminate := false;
  Canceled := false;
  Xls := aXls;
  ProgressFeedback := aProgressFeedback;
  FinalFeedback := aFinalFeedback;
  FileName := aFileName;
  AllSheets := aAllSheets;
end;

procedure TProgressThread.DoTerminate;
var
  msg: string;
  ok: boolean;
begin
  inherited;

  msg := '';
  ok := not Canceled;
  if (Assigned(FatalException)) then
  begin
    ok := false;
    if FatalException is Exception then
    begin
      msg := Exception(FatalException).Message;
    end
    else msg := 'Unexpected error: ' + FatalException.ClassName;
  end;

  if Canceled then msg := 'Operation canceled by the user.';


  //DoTerminate runs in the thread context, not the main thread context as OnTerminate.
  Synchronize(
  procedure
  begin
    FinalFeedback(ok, msg);
  end);
end;

{$IF CompilerVersion > 22} //Delphi XE doesn't support TerminatedSet, so sadly in XE we will have to call it manually.
procedure TProgressThread.TerminatedSet;
begin
  inherited;
  Canceled := true;
end;
{$else}
procedure TProgressThread.Cancel;  //this isn't nice, because we need to remember to call cancel instead of Terminate for our thread. But in XE there is no option.
begin
  Canceled := true;
end;
{$ifend}
end.

File: Cpp/Modules/25.Printing and Exporting/20.CustomPreview/UProgressThread.cpp
//---------------------------------------------------------------------------

#pragma hdrstop

#include "UProgressThread.h"
//---------------------------------------------------------------------------
#pragma package(smart_init)

__fastcall TProgressThread::TProgressThread(TExcelFile *aXls, TFeedbackMethod aProgressFeedback,
	  TFinalFeedbackMethod aFinalFeedback, String aFileName, bool aAllSheets): TThread(true)
{
  FreeOnTerminate = false;
  Canceled = false;
  Xls = aXls;
  ProgressFeedback = aProgressFeedback;
  FinalFeedback = aFinalFeedback;
  FileName = aFileName;
  AllSheets = aAllSheets;

}

void __fastcall TProgressThread::TerminatedSet()
{
  TThread::TerminatedSet();
  Canceled = true;
}
//---------------------------------------------------------------------------

void __fastcall TProgressThread::DoTerminate()
{
  TThread::DoTerminate();
  FinalFeedback_msg = "";
  FinalFeedback_ok = !Canceled;
  if (FatalException != NULL)
  {
	FinalFeedback_ok = false;
	if (dynamic_cast<Exception*>(FatalException))
	{
	  FinalFeedback_msg = ((Exception*)FatalException)->Message;
	}
	else FinalFeedback_msg = "Unexpected error: " + FatalException->ClassName();
  }

  if (Canceled) FinalFeedback_msg = "Operation canceled by the user.";


  //DoTerminate runs in the thread context, not the main thread context as OnTerminate.
  Synchronize(&DoFinalFeedback);

}
//---------------------------------------------------------------------------

void __fastcall TProgressThread::DoFinalFeedback()
{
	FinalFeedback(FinalFeedback_ok, FinalFeedback_msg);
}

void __fastcall TProgressThread::DoProgressFeedback()
{
	ProgressFeedback(ProgressFeedback_percent, ProgressFeedback_msg);
}

